drivers/memc/memc_nxp_s32_qspi.c

```diff
@@ -1,197 +1,1707 @@
+	uint32_t lock_mask =
+		slave ? QSPI_DLLSR_SLV_LOCK_MASK(port) : QSPI_DLLSR_DLL_LOCK_MASK(port);
+	uint32_t err_mask =
+		QSPI_DLLSR_DLL_RANGE_ERR_MASK(port) | QSPI_DLLSR_DLL_FINE_UNDERFLOW_MASK(port);
+	bool locked = false;
+	bool error = false;
+
+	while (!locked && !error && (cycle_count > (k_cycle_get_32() - cycle_start))) {
+		locked = FIELD_GET(lock_mask, REG_READ(QSPI_DLLSR)) != 0U;
+		error = FIELD_GET(err_mask, REG_READ(QSPI_DLLSR)) != 0U;
+	}
+
+	return error ? -EIO : (locked ? 0 : -ETIMEDOUT);
+}
+
+static int qspi_dll_config_bypass(const struct qspi_config *config, uint8_t port)
+{
+	const struct qspi_controller *controller = config->controller;
+	uint32_t reg_val;
+	int ret = 0;
+
+	/* Set DLL in bypass mode */
+	reg_val = REG_READ(QSPI_DLLCR(port));
+	reg_val |= QSPI_DLLCR_SLV_DLL_BYPASS(1U);
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_SLAVE_AUTO_UPDT)
+	reg_val &= ~QSPI_DLLCR_SLAVE_AUTO_UPDT_MASK;
+#endif /* CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_SLAVE_AUTO_UPDT */
+	REG_WRITE(QSPI_DLLCR(port), reg_val);
+
+	/* Program desired DQS delay for sampling */
+	reg_val = REG_READ(QSPI_DLLCR(port));
+	reg_val &= ~(QSPI_DLLCR_SLV_DLY_COARSE_MASK | QSPI_DLLCR_SLV_FINE_OFFSET_MASK |
+		     QSPI_DLLCR_FREQEN_MASK);
+	reg_val |= QSPI_DLLCR_SLV_DLY_COARSE(controller->ports[port].dll.coarse_delay);
+	reg_val |= QSPI_DLLCR_SLV_FINE_OFFSET(controller->ports[port].dll.fine_delay);
+	reg_val |= QSPI_DLLCR_FREQEN(controller->ports[port].dll.freq_enable);
+	REG_WRITE(QSPI_DLLCR(port), reg_val);
+
+	/* Trigger update to load these values in the slave delay chain */
+	qspi_dll_slave_update(config, port, true);
+
+	/* Wait for slave delay chain update to take effect */
+	ret = qspi_dll_wait_lock(config, port, true);
+
+	/* Disable slave chain update */
+	qspi_dll_slave_update(config, port, false);
+
+	return ret;
+}
+
+static int qspi_dll_config_update(const struct qspi_config *config, uint8_t port)
+{
+	const struct qspi_controller *controller = config->controller;
+	uint32_t reg_val;
+	int ret = 0;
+
+	/* Set DLL in auto update mode */
+	reg_val = REG_READ(QSPI_DLLCR(port));
+	reg_val &= ~QSPI_DLLCR_SLV_DLL_BYPASS_MASK;
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_SLAVE_AUTO_UPDT)
+	if (controller->ports[port].dll.mode == QSPI_DLL_AUTO_UPDATE) {
+		reg_val |= QSPI_DLLCR_SLAVE_AUTO_UPDT(1U);
+	} else {
+		reg_val &= ~QSPI_DLLCR_SLAVE_AUTO_UPDT_MASK;
+	}
+#endif /* CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_SLAVE_AUTO_UPDT */
+	REG_WRITE(QSPI_DLLCR(port), reg_val);
+
+	/* Program DLL configuration */
+	reg_val = REG_READ(QSPI_DLLCR(port));
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_REFCNTR)
+	reg_val &= ~QSPI_DLLCR_DLL_REFCNTR_MASK;
+	reg_val |= QSPI_DLLCR_DLL_REFCNTR(controller->ports[port].dll.reference_counter);
+#endif /* CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_REFCNTR */
+#if defined(CONFIG_MEMC_NXP_S32_QSPI_HAS_DLLCR_DLLRES)
+	reg_val &= ~QSPI_DLLCR_DLL_DLLRES_MASK;
+	reg_val |= QSPI_DLLCR_DLL_DLLRES(controller->ports[port].dll.resolution);
```

Comment:
these macros `QSPI_DLLCR_DLL_DLLRES_MASK` and `QSPI_DLLCR_DLL_DLLRES` are not defined